hail/src/main/scala/is/hail/expr/ir/Infer.scala

```diff
@@ -61,15 +61,7 @@ object Infer {
         val tbs = coerce[TStruct](old.typ)
         TStruct(fields.map { id => (id, tbs.field(id).typ) }: _*)
       case InsertFields(old, fields) =>
-        fields.foldLeft(old.typ) { case (t, (name, a)) =>
-          t match {
-            case t2: TStruct =>
-              t2.selfField(name) match {
-                case Some(f2) => t2.updateKey(name, f2.index, a.typ)
-                case None => t2.appendKey(name, a.typ)
-              }
-          }
-        }.asInstanceOf[TStruct]
+        old.typ.asInstanceOf[TStruct].insertFields(fields.map(x => (x._1, x._2.typ)))
```

Comment:
I think the solution here is to assert that InsertFields doesn't have an overlap between the old and new, and drop stuff in python when it appears. This will simplify logic elsewhere (e.g. PruneDeadFields). But this is also way better

Reply:
Yeah, unless you have specific objections to these changes, I say we put them in.

Reply:
oh right @patrick-schultz was the reviewer, never mind!